### PR TITLE
refactor: make tr_torrent idle fields private

### PR DESCRIPTION
**Pull request by ckerr**

### CODE REVIEW

File: libtransmission/torrent.cc [hunk 1]
```diff
@@ -1497,6 +1428,8 @@ tr_stat const* tr_torrentStat(tr_torrent* tor)
     auto seed_ratio_bytes_goal = uint64_t{};
     bool const seed_ratio_applies = tr_torrentGetSeedRatioBytes(tor, &seed_ratio_bytes_left, &seed_ratio_bytes_goal);
 
+    s->eta = TR_ETA_NOT_AVAIL;
+    s->etaIdle = TR_ETA_NOT_AVAIL;
```
Comment (tearfur) [inline libtransmission/torrent.cc:1432]: These lines can go:

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1450

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1461

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1500-L1501

These lines can be condensed:

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1464-L1469

```cpp
    case TR_STATUS_SEED:
        if (seed_ratio_applies)
```

File: libtransmission/torrent.h [hunk 2]
```diff
@@ -711,39 +711,114 @@ struct tr_torrent final : public tr_completion::torrent_view
...
+    [[nodiscard]] constexpr auto peer_limit() const noexcept
+    {
+        return max_connected_peers_;
+    }
+
+    // --- idleness
+
+    constexpr void set_idle_limit_mode(tr_idlelimit mode) noexcept
+    {
+        if (mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED)
```
Comment (tearfur) [inline libtransmission/torrent.h:723]: Maybe adding back an assert is a good idea.

```cpp
TR_ASSERT(mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED);
```

File: libtransmission/torrent.cc (hunk 1 above)
Comment (tearfur) [inline libtransmission/torrent.cc:1432]: These lines can go:

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1450

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1461

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1500-L1501

These lines can be condensed:

https://github.com/transmission/transmission/blob/24068f75d72262e0e585fcb0d168baf49d000e77/libtransmission/torrent.cc#L1464-L1469

```cpp
    case TR_STATUS_SEED:
        if (seed_ratio_applies)
```

File: libtransmission/torrent.h (hunk 2 above)
Comment (tearfur) [inline libtransmission/torrent.h:723]: Maybe adding back an assert is a good idea, we don't want to write bugs and have libtransmission silently ignore them for us.

```cpp
TR_ASSERT(mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED);
```

File: libtransmission/torrent.h [hunk 3]
```diff
@@ -711,39 +711,114 @@ struct tr_torrent final : public tr_completion::torrent_view
...
+        return idle_limit_seconds > *idle_seconds ? idle_limit_seconds - *idle_seconds : 0U;
     }
 
-    constexpr void set_ratio_mode(tr_ratiolimit mode) noexcept
+    // --- seed ratio
+
+    constexpr void set_seed_ratio_mode(tr_ratiolimit mode) noexcept
     {
-        if (ratioLimitMode != mode)
+        if (mode == TR_RATIOLIMIT_GLOBAL || mode == TR_RATIOLIMIT_SINGLE || mode == TR_RATIOLIMIT_UNLIMITED)
```
Comment (tearfur) [inline libtransmission/torrent.h:774]: ditto

```cpp
TR_ASSERT(mode == TR_IDLELIMIT_GLOBAL || mode == TR_IDLELIMIT_SINGLE || mode == TR_IDLELIMIT_UNLIMITED);
```

File: libtransmission/torrent.h [hunk 4]
```diff
@@ -932,6 +1000,38 @@ struct tr_torrent final : public tr_completion::torrent_view
...
+        return {};
+    }
+
+    [[nodiscard]] constexpr std::optional<size_t> idle_seconds(time_t now) const noexcept
+    {
+        auto const activity = this->activity();
+
+        if (activity == TR_STATUS_DOWNLOAD || activity == TR_STATUS_SEED)
+        {
+            if (auto const latest = std::max(startDate, activityDate); latest != 0 && now >= latest)
```
Comment (tearfur) [inline libtransmission/torrent.h:1026]: Could be a good idea to assert `now >= latest` here. I can't think of any use case where `now` isn't `tr_time()` at the moment.

**Review by tearfur — COMMENTED** [review]
I'll be happy to merge as-is. But just some minor nits.

File: libtransmission/torrent.cc (hunk 1 above)
Comment (ckerr) [reply libtransmission/torrent.cc:1433]: updated in 713920e08036c5b0ef307f7a03d68b30d8fe5908

File: libtransmission/torrent.h (hunk 3 above)
Comment (ckerr) [reply libtransmission/torrent.h]: restored in 4e07e7bfcca6e0896d2270ccc8e9f14edd1f4c14

File: libtransmission/torrent.h (hunk 4 above)
Comment (ckerr) [reply libtransmission/torrent.h]: good call; added in 9cfa102764f2366368485c5e5e087112660bd395

File: libtransmission/torrent.cc [hunk 5]
```diff
@@ -1512,28 +1445,18 @@ tr_stat const* tr_torrentStat(tr_torrent* tor)
...
-        {
-            s->eta = TR_ETA_NOT_AVAIL;
-        }
-        else if (tor->etaSpeed_Bps == 0)
+        if (tor->etaSpeed_Bps == 0)
         {
             s->eta = TR_ETA_UNKNOWN;
         }
-        else
+        else if (s->leftUntilDone <= s->desiredAvailable)
```
Comment (tearfur) [inline libtransmission/torrent.cc:1452]:
```suggestion
        else if (s->leftUntilDone <= s->desiredAvailable || tor->webseed_count() >= 1U)
```

**Comment by tearfur** [comment]
@ckerr I'd like to, but I can't. 😅

I don't have the resolve button in your PRs.

**Comment by ckerr** [comment]
Hmm! Okay. I'll need to figure out how to fix that.

In the interim, maybe re-review with an Approve when you're happy? Same rationale: when I make changes to address review feedback, I don't like to merge until the reviewer's had a chance to look over the newer changes.

**Pull request merged**